protobuf/sync/v1/sync_service.proto

```diff
@@ -0,0 +1,52 @@
+syntax = "proto3";
+
+package sync.v1;
+
+option go_package = "flagd/grpcsync";
+
+// SyncFlagsRequest is the request initiating the sever-streaming rpc. Flagd sends this request, acting as the client
+message SyncFlagsRequest {
+  // Optional: A unique identifier for flagd  provider (grpc client) initiating the request. The server implementations
+  // can utilize this identifier to aggregate flag configurations and stream them to a specific client. This identifier
+  // is intended to be optional. However server implementation may enforce it.
+  string provider_id = 1;
+}
+
+// SyncState convey the state of the payload. These states are related to flagd isync.go type definitions but
+// contains extras to optimize grpc use case. Refer - https://github.com/open-feature/flagd/blob/main/pkg/sync/isync.go
+enum SyncState {
+  // Value is ignored by the listening flagd
+  SYNC_STATE_UNSPECIFIED = 0;
+
+  // All the flags matching the request. This is the default response and other states can be ignored
+  // by the implementation. Flagd internally replaces all existing flags for this response state.
+  SYNC_STATE_ALL = 1;
+
+  // Convey an addition of a flag. Flagd internally handles this by combining new flags with existing ones
+  SYNC_STATE_ADD = 2;
+
+  // Convey an update of a flag. Flagd internally attempts to update if the updated flag already exist OR if it does not,
+  // it will get added
+  SYNC_STATE_UPDATE = 3;
+
+  // Convey a deletion of a flag. Flagd internally removes the flag
+  SYNC_STATE_DELETE = 4;
+
+  // Optional server ping to check client connectivity. Handling is ignored by flagd and is to merely support live check
+  SYNC_STATE_PING = 5;
+}
+
+// SyncFlagsResponse is the sever response containing feature flag configurations and the state
+message SyncFlagsResponse {
+  // flagd feature flag configuration
+  string flags = 1;
```

Comment:
I'm not sure `flags` is much clearer.

I think we need to choose a name that doesn't collide with other identifiers (we already use "flags" in the flagd schema).

Reply:
Actually - might this need to be a string array?

When flagd connects, in the case of OFO, for example, it might pass multiple `featureFlagConfiguration` CR names it's interested in. I'm not sure we want to create a new abstraction that combines these... unless we can use the existing `flagSourceConfiguration` CR for this purpose.

Reply:
I took the approach to have simple, valid flagd feature flag configurations in json format here (matching our schema definition [1]). Yes, naming can be imrpvoed: `flag_configurations` instead of `flags` and referring to schema makes it clearer. 

Regarding string vs array argument, I am not sure whether we should explore that direction now 🤔 The current implementation only focuses on one sync source and that source provide single flagd configuration payload. 

Related conversion - https://github.com/open-feature/schemas/pull/78#discussion_r1097772791

Anyway I am open for more discussion here :)

[1]. - https://raw.githubusercontent.com/open-feature/schemas/main/json/flagd-definitions.json

Reply:
One of our main use cases is to use this in OFO - so I'd like to make sure we aren't paining ourselves into a corner with this schema change.

A flagd sidecare process might be interested in multiple CRs in the OFO paradigm. I think, given this, we _need_ a means for that sidecar flagd to tell the server which CRs it cares about when it initiates the gRPC connection. From where I'm standing that gives us a few options:

- it references a single `flagSourceConfiguration` that (with @james-milligan new incoming changes) will reference mulitple `featureFlagConfigurations`
- it references multiple `featureFlagConfiguration` directly in an array.
- it starts multiple gRPC syncs, each caring about a different `featureFlagConfiguration`

Reply:
Should we really consider grpc server implementation internals or should we only consider as a single source? Per my understanding, this is the concern isn't it ? 

Per current implementation, a start-up can provide multiple sync providers, each focusing on an individual sync source. GRPC sync, in my view, is no different and consider the stream as from a single source. 

I am yet to see OFO use-case though. @toddbaert  may be can you come up with a sketch to visualize your concern/proposal ?

Reply:
> Per current implementation, a start-up can provide multiple sync providers, each focusing on an individual sync source. GRPC sync, in my view, is no different and consider the stream as from a single source.

I'm 100% fine with that, it's should work just fine, I suppose, and maybe keeps things simple. I mentioned "it starts multiple gRPC syncs, each caring about a different featureFlagConfiguration" as a valid option, so I agree on that.

> I am yet to see OFO use-case though. @toddbaert may be can you come up with a sketch to visualize your concern/proposal ?

I think the only thing I'm saying is if we wanted, we could make the gRPC stream aware of multiple sources, which would get us some efficiency (perhaps)? But it may be overly complex.

Reply:
The complexity comes with aligning such sync mechanism to existing isync interface [1]. Current isync interface is simple and see sync provider with one to one basis (one source, one sync provider). But yes, there can be use cases for one to many (one source, multiple underlying sources). 

My view is that,

 **first** strengthen the simple use-case: Ofep, schema, simple sync and security (I call it the core)

**secondly**, introduce a multi-source sync method: non-breaking, backward-compatible change with a new rpc method (`SyncFlags` vs `SyncFlagsMutiSource`), where servers can opt-in to support


What do you think ?  

cc - @james-milligan may be this maps to your idea at comment https://github.com/open-feature/schemas/pull/78#discussion_r1097772791



[1] - https://github.com/open-feature/flagd/blob/main/pkg/sync/isync.go

Reply:
I like this approach